raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -580,9 +581,16 @@ private EndQuorumEpochResponseData handleEndQuorumEpochRequest(
             if (state.isUnattached()
                 || state.hasLeader(requestReplicaId)
                 || state.hasVotedFor(requestReplicaId)) {
-                // reset the timer to election jitter so that they would become candidate
-                // after a pretty short period of time
-                timer.reset(randomElectionJitterMs());
+                List<Integer> preferredSuccessors = request.preferredSuccessors();
+                // We didn't find the corresponding voters inside the request.
+                if (!preferredSuccessors.contains(quorum.localId)) {
+                    return buildEndQuorumEpochResponse(Errors.INCONSISTENT_VOTER_SET);
+                }
+                final int position = preferredSuccessors.indexOf(quorum.localId);
+                // Based on the priority inside the preferred successors, choose the corresponding delayed
+                // election time so that the most up-to-date voter has a higher chance to be elected.
+                timer.reset(position <= 0 ? 0 :
```

Comment:
Hmm.. a tricky thing here is that we would only check the timer again in the next poll call so the first preferred candidate would not try to become a candidate right away still, and if it calls the next poll after more than `retryBackoffMs` it would actually be in the same starting line as the second preferred candidate.

So maybe we should just execute the `becomeCandidate` directly than setting a timer to 0 here.

Reply:
I see, makes sense